### DISCUSSION TITLE
How do I get the limited API version from `wheel.py-api` into `USE_SABI`?

### DISCUSSION THREAD

**Discussion by minrk** [scikit-build/scikit-build-core#1470] [Q&A]
I am exploring limited API wheel suport in the Cython 3.1 alpha, and I am trying to figure out the simplest way to set the right API version to match the wheel tag. There are valuable features (I think) for my case in the 3.11 and 3.12 Limited APIs, so I am considering building one 3.8 wheel and at least one or both of 3.11 and 3.12.

I expected setting `wheel.pyapi=cp38` to set the equivalent of `USE_SABI 3.8`, but it doesn't (I'm sure there's a good reason). And I can't seem to find the `cp38` string anywhere in the cmake variables. What is the simplest way to set the SABI to match the version specified in wheel.py-api and/or set both from the same place?

**Comment by minrk** [comment]
For more detail, I'm exploring building limited API wheels for pyzmq with Cython 3.1 and cibuildwheel in https://github.com/zeromq/pyzmq/pull/2051

A key point is that I can't just switch to a single stable api wheel and be done, because I need the buffer APIs added to the stable API in 3.11, but still support older Python than that. So I need to at least build wheels for <=3.10, 313t, pypy, and then a stable wheel for cp311. I'm trying to figure out the best way to do that where I want _some_ stable API wheels targeting specific versions and other regular wheels in the most maintainable way I can.

**Reply by jcfr** [reply]
When setting `wheel.py-api`[^1][^2], the CMake variable `SKBUILD_SABI_COMPONENT` is set to `Development.SABIModule`[^3] so that it can be conveniently used when calling `find_package(Python ...). For example:

```cmake
find_package(Python REQUIRED COMPONENTS Interpreter Development.Module ${SKBUILD_SABI_COMPONENT})`
```

Improving scikit-build-core so that it also conveniently set a variable called `SKBUILD_SABI_VERSION` set that it can be used when calling `Python3_add_library` would be sensible. That way we could do the following:

```cmake
if(NOT "${SKBUILD_SABI_COMPONENT}" STREQUAL "")
  python_add_library(some_ext MODULE WITH_SOABI USE_SABI ${SKBUILD_SABI_VERSION} ...)
else()
  python_add_library(some_ext MODULE WITH_SOABI ...)
endif()
```

In the meantime, the approach you implemented in the pull request linked above is sensible.

Related issues & pull-requests:
* https://github.com/scikit-build/scikit-build-core/issues/958
* https://github.com/scikit-build/scikit-build-core/issues/743
* https://github.com/scikit-build/scikit-build-core/pull/957

cc: @henryiii 

[^1]: https://scikit-build-core.readthedocs.io/en/latest/cmakelists.html#limited-api-stable-abi
[^2]: https://scikit-build-core.readthedocs.io/en/latest/configuration.html#customizing-the-output-wheel
[^3]: https://cmake.org/cmake/help/latest/module/FindPython3.html#findpython3

**Reply by minrk** [reply]
Yeah, my approach _assumes_ that if building with the stable ABI, the wheel tag is the build-time Python's version. But that's not necessarily the case, and can easily result in invalid wheels without any checks. I think https://github.com/scikit-build/scikit-build-core/issues/958 would be a great help to make sure the output matches the input.

It would probably also be nice if I could specify with a _boolean_ flag to use the stable ABI with the build-time Python as the target. I think that would help integration with cibuildhweel, so I wouldn't need a separate `overrides` entry for every combination, only one expressing when stable abi wheels should be built.
